### PR TITLE
Improve completion logic

### PR DESCRIPTION
**Pull request by andschwa**
This makes the deduced `CompletionItemKind` enum more accurate, which
results in better icons for the user when viewing completions in an
editor. Also cleaned up the file as there were remnants of dead code.

Fixes https://github.com/PowerShell/vscode-powershell/issues/3364

### CODE REVIEW

**Comment by andschwa** [comment]
@MartinGC94 I think you'll like this!

From your demo (ignore `ProviderItem`, Code does some magic to get filetype specific icons later):
Screen Shot 2022-03-07 at 4 42 07 PM

Files:
Screen Shot 2022-03-07 at 4 44 43 PM

Operators:
Screen Shot 2022-03-07 at 4 46 06 PM

Namespace:
Screen Shot 2022-03-07 at 4 43 38 PM

And even classes:
Screen Shot 2022-03-07 at 4 43 09 PM

**Comment by MartinGC94** [comment]
You are right, this looks great and great choice of icons as well. The only icon that looks a little out of place is the ParameterName. I know it shares its icon with the operators but parameters are used much more frequently so maybe the chosen icon should look more like a parameter?

File: src/PowerShellEditorServices/Services/TextDocument/Handlers/CompletionHandler.cs [hunk 1]
```diff
@@ -110,13 +104,9 @@ public async Task<CompletionItem> Handle(CompletionItem request, CancellationTok
             }
 
             // Get the documentation for the function
-            CommandInfo commandInfo =
-                await CommandHelpers.GetCommandInfoAsync(
-                    request.Label,
-                    _runspaceContext.CurrentRunspace,
-                    _executionService).ConfigureAwait(false);
+            CommandInfo commandInfo = await CommandHelpers.GetCommandInfoAsync( request.Label, _runspaceContext.CurrentRunspace, _executionService).ConfigureAwait(false);
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Services/TextDocument/Handlers/CompletionHandler.cs:107]:
```suggestion
            CommandInfo commandInfo = await CommandHelpers.GetCommandInfoAsync(request.Label, _runspaceContext.CurrentRunspace, _executionService).ConfigureAwait(false);
```

File: src/PowerShellEditorServices/Services/TextDocument/Handlers/CompletionHandler.cs [hunk 2]
```diff
@@ -323,43 +304,33 @@ public void SetCapability(CompletionCapability capability, ClientCapabilities cl
...
+                CompletionType.Property => CompletionItemKind.Property,
+                CompletionType.Variable => CompletionItemKind.Variable,
+                CompletionType.Namespace => CompletionItemKind.Module,
+                CompletionType.Type => CompletionItemKind.TypeParameter,
+                CompletionType.Keyword => CompletionItemKind.Keyword,
+                CompletionType.File => CompletionItemKind.File,
+                CompletionType.Folder => CompletionItemKind.Folder,
+                CompletionType.History => CompletionItemKind.Reference,
+                CompletionType.Text => CompletionItemKind.Text,
+                _ => throw new NotImplementedException(),
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Services/TextDocument/Handlers/CompletionHandler.cs:327]:
```suggestion
                _ => throw new ArgumentOutOfRangeException(nameof(completionType)),
```

`NotImplementedException` is usually for interface implementations that are explicitly not implemented.

**Review by SeeminglyScience — APPROVED** [review]
Looks great! Just a couple of nits

**Comment by andschwa** [comment]
> You are right, this looks great and great choice of icons as well. The only icon that looks a little out of place is the ParameterName. I know it shares its icon with the operators but parameters are used much more frequently so maybe the chosen icon should look more like a parameter?

@MartinGC94 That is an artifact of the way that test displays all of them at once and the logic that figures out if it's a parameter. They're actually currently mapped to `CompletionItemKind.Variable` (which was their previous mapping) and look like this:
Screen Shot 2022-03-08 at 9 47 05 AM

However, I'm trying to decide if instead they should map to `Field` or `Property` (which are maybe slightly more accurate? The parameters declared on a PowerShell function are kind of like fields/properties).

Field (which is so far unused):
Screen Shot 2022-03-08 at 9 50 11 AM

Property (which already has the PowerShell `CompletionResultType.Property` mapping to it):
Screen Shot 2022-03-08 at 9 51 57 AM

**Comment by MartinGC94** [comment]
I see. All 3 of those are fine but if I had to pick I would go with Field because I like the icon and the fact that it's unique.

**Comment by JustinGrote** [comment]
Using other languages as prior art, parameters are generally considered variables (Typescript as example):

image
image

C# as another:
image


So I think it should stay as variable for a consistent experience.

**Comment by andschwa** [comment]
Thanks for pointing that out @JustinGrote, I agree. What tool are you using to inspect like that?

**Comment by JustinGrote** [comment]
@andschwa it's built into vscode, very useful for troubleshooting syntax highlighting
image

**Comment by andschwa** [comment]
> @andschwa it's built into vscode, very useful for troubleshooting syntax highlighting
> image

I'm so sad I've learned about this a year later...what else should I know?!

**Comment by JustinGrote** [comment]
@andschwa uuuh, you can access devtools for the running vscode instance maybe?
image

Basically check out everything prefixed with `Developer:`

**Comment by andschwa** [comment]
@JustinGrote @SeeminglyScience @MartinGC94 Give this new commit a whirl would you? I removed the middle abstraction layer as it seemed superfluous.

**Comment by andschwa** [comment]
Oops, probably tests expecting old code. Fixing up.

File: src/PowerShellEditorServices/Services/TextDocument/Handlers/CompletionHandler.cs [hunk 3]
```diff
@@ -97,34 +80,32 @@ public static bool CanResolve(CompletionItem value)
...
-            {
-                return request;
-            }
-
-            // No details means the module hasn't been imported yet and Intellisense shouldn't import the module to get this info.
-            if (request.Detail is null)
+            // We currently only support this request for anything that returns a CommandInfo:
+            // functions, cmdlets, aliases. No detail means the module hasn't been imported yet and
+            // IntelliSense shouldn't import the module to get this info.
+            if (request.Kind != CompletionItemKind.Function || request.Detail is null)
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Services/TextDocument/Handlers/CompletionHandler.cs:86]:
```suggestion
            if (request.Kind is not CompletionItemKind.Function || request.Detail is null)
```

nit: `is` can be used whenever the RHS is a constant

File: src/PowerShellEditorServices/Services/TextDocument/Handlers/CompletionHandler.cs [hunk 4]
```diff
@@ -149,217 +130,170 @@ public void SetCapability(CompletionCapability capability, ClientCapabilities cl
...
-        private static CompletionItemKind MapCompletionKind(CompletionType completionType)
+        /// <summary>
+        /// Look for type encoded in the tooltip for parameters and variables. Display PowerShell
+        /// type names in [] to be consistent with PowerShell syntax and how the debugger displays
+        /// type names.
+        /// </summary>
+        /// <param name="toolTipText"></param>
+        /// <param name="type"></param>
+        /// <returns>Whether or not the type was found.</returns>
+        private static bool ExtractTypeFromToolTip(string toolTipText, out string type)
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Services/TextDocument/Handlers/CompletionHandler.cs:256]: nit: For this pattern usually the convention in C# is to name it `TryEtc`

File: src/PowerShellEditorServices/Services/TextDocument/Handlers/CompletionHandler.cs [hunk 5]
```diff
@@ -149,217 +130,170 @@ public void SetCapability(CompletionCapability capability, ClientCapabilities cl
...
         {
-            switch (completionType)
+            Regex _typeRegex = new(@"^(\[.+\])");
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Services/TextDocument/Handlers/CompletionHandler.cs:258]:
```suggestion
            Regex typeRegex = new(@"^(\[.+\])");
```

For locals skip `_` (I know it's to mirror the real field but still, better to just reference the new local)

File: src/PowerShellEditorServices/Services/TextDocument/Handlers/CompletionHandler.cs [hunk 6]
```diff
@@ -149,217 +130,170 @@ public void SetCapability(CompletionCapability capability, ClientCapabilities cl
...
+            MatchCollection matches = _typeRegex.Matches(toolTipText);
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Services/TextDocument/Handlers/CompletionHandler.cs:259]:
```suggestion
            MatchCollection matches = typeRegex.Matches(toolTipText);
```

**Review by SeeminglyScience — APPROVED** [review]
LGTM, couple of little things

**Pull request merged**